python/pylibcudf/pylibcudf/scalar.pyx

```diff
@@ -152,53 +155,107 @@ cdef Scalar _new_scalar(unique_ptr[scalar] c_obj, DataType dtype):
 
 
 @singledispatch
-def _from_py(py_val):
+def _from_py(py_val, dtype: DataType | None):
     raise TypeError(f"{type(py_val).__name__} cannot be converted to pylibcudf.Scalar")
 
 
 @_from_py.register(dict)
 @_from_py.register(list)
 @_from_py.register(datetime.datetime)
 @_from_py.register(datetime.timedelta)
-def _(py_val):
+def _(py_val, dtype: DataType | None):
     raise NotImplementedError(
         f"Conversion from {type(py_val).__name__} is currently not supported."
     )
 
 
 @_from_py.register(float)
-def _(py_val):
-    cdef DataType dtype = DataType(type_id.FLOAT64)
-    cdef unique_ptr[scalar] c_obj = make_numeric_scalar(dtype.c_obj)
-    (<numeric_scalar[double]*>c_obj.get()).set_value(py_val)
-    cdef Scalar slr = _new_scalar(move(c_obj), dtype)
-    return slr
+def _(py_val: float, dtype: DataType | None):
+    if dtype is None:
+        dtype = DataType(type_id.FLOAT64)
+
+    cdef unique_ptr[scalar] c_obj
+    tid = dtype.id()
+
+    if tid == type_id.FLOAT32:
+        c_obj = make_numeric_scalar((<DataType>dtype).c_obj)
+        (<numeric_scalar[float]*>c_obj.get()).set_value(py_val)
```

Comment:
Curious what happens if we make a float32 scalar but the `py_val` is greater than the max float32 value (similar to the integer cases below).

Reply:
We might need to raise in those cases